packages/markuplint/src/resolve-rules.ts

```diff
@@ -7,7 +7,7 @@ export async function resolveRules(options: { rules?: MLRule<RuleConfigValue, un
 		rules = options.rules;
 	} else {
 		const r = await import('@markuplint/rules');
-		rules = r.default;
+		rules = r.default as MLRule<RuleConfigValue, unknown>[];
```

Comment:
@JounQin 

Why need it? `rules`'s type is already defined.

Reply:
I have no idea too.

![image](https://user-images.githubusercontent.com/8336744/116365499-e1ce6680-a837-11eb-90ae-55ff1326387a.png)

```log
src/resolve-rules.ts(10,3): error TS2322: Type '(MLRule<null, null> | MLRule<"always" | "never" | "always-single-line" | "never-single-line", null> | MLRule<boolean, AttrSpasingOptions> | ... 12 more ... | MLRule<...>)[]' is not assignable to type 'MLRule<RuleConfigValue, unknown>[]'.
  Type 'MLRule<null, null> | MLRule<"always" | "never" | "always-single-line" | "never-single-line", null> | MLRule<boolean, AttrSpasingOptions> | ... 12 more ... | MLRule<...>' is not assignable to type 'MLRule<RuleConfigValue, unknown>'.
    Type 'MLRule<null, null>' is not assignable to type 'MLRule<RuleConfigValue, unknown>'.
      Types of property 'v' are incompatible.
        Type '(document: MLDOMDocument<null, null>, translate: Translator, globalRule: RuleInfo<null, null>) => Result[] | Promise<Result[]>' is not assignable to type '(document: MLDOMDocument<RuleConfigValue, unknown>, translate: Translator, globalRule: RuleInfo<RuleConfigValue, unknown>) => Result[] | Promise<...>'.
          Types of parameters 'document' and 'document' are incompatible.
            Type 'MLDOMDocument<RuleConfigValue, unknown>' is not assignable to type 'MLDOMDocument<null, null>'.
              Types of property 'nodeList' are incompatible.
                Type 'readonly AnonymousNode<RuleConfigValue, unknown>[]' is not assignable to type 'readonly AnonymousNode<null, null>[]'.
                  Type 'AnonymousNode<RuleConfigValue, unknown>' is not assignable to type 'AnonymousNode<null, null>'.
                    Type 'MLDOMDoctype<RuleConfigValue, unknown>' is not assignable to type 'AnonymousNode<null, null>'.
                      Type 'MLDOMDoctype<RuleConfigValue, unknown>' is not assignable to type 'MLDOMDoctype<null, null>'.
                        Type 'RuleConfigValue' is not assignable to type 'null'.
                          Type 'string' is not assignable to type 'null'.
```

Reply:
Oh..., Ok.

Reply:
@YusukeHirao 

This should be related to the added two getters.

There is a tick that we can mark the two getters `private`, then there will be no problem here.

But the subclass must use `// @ts-ignore` to suspend the ts error for accessing `private` getter.

This tick will work on runtime, because ts `private` keyword is `private` for ts code only.

However, I don't know should we do like so. It makes these two getters not `public` API.

Reply:
Uh‐huh, Umm...
Well...

Actually, I don't run on my computer yet.
I will see the details tonight.
Please wait for.

Reply:
I decide to use `private` pattern to make sure the types are compatible with before.

And I've already used `// @ts-ignore` pattern at `markuplint-sync`, because the `constructor` of `MLRule` is already `private`.

See https://github.com/rx-ts/markuplint-sync/pull/1/files#diff-f1dab096c58accb446e00e4dc82e2da5676fe57671a8b2125fc2b88bf0a610deR10

Reply:
Oh, right? If that is what you'd like I'm ok.
(However, that type inference was strange...🤔 )

Reply:
@YusukeHirao 

> However, that type inference was strange...🤔

Every rule has its own `RuleConfigValue` and `Option`, and the two getters points to `#v/#f` but public, so they are not compatible.

Reply:
@JounQin 

It's clear!

Please change them like below:

```ts
/**
 * The following getter is unused internally,
 * only for extending from 3rd party library
 */
protected get v(): MLRuleOptions<T, O>['verify'] {
	return this.#v;
}

/**
 * The following getter is unused internally,
 * only for extending from 3rd party library
 */
protected get f(): MLRuleOptions<T, O>['fix'] {
	return this.#f;
}
```

The point is generic types `T` and `O`.
And I thought about that again, the comment says "for 3rd party" so the accessor is `private` is strange. So I want to change it to `protected`.

Reply:
Wow, I was thinking `#v` is already `MLRuleOptions<T, O>['verify']`. A strange behavior of TypeScript to me.